src/openMVG/sfm/sfm_data_graph_utils.cpp

```diff
@@ -0,0 +1,148 @@
+// This file is part of OpenMVG, an Open Multiple View Geometry C++ library.
+
+// Copyright (c) 2015, 2017 WhuAegeanSea, Pierre Moulon.
+
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "openMVG/sfm/sfm_data_graph_utils.hpp"
+
+#include "openMVG/graph/connectedComponent.hpp"
+#include "openMVG/graph/graph.hpp"
+#include "openMVG/graph/graph_builder.hpp"
+#include "openMVG/sfm/sfm_data.hpp"
+#include "openMVG/sfm/pipelines/sfm_matches_provider.hpp"
+#include "openMVG/stl/stl.hpp"
+#include "openMVG/types.hpp"
+
+#include <fstream>
+#include <iosfwd>
+
+namespace openMVG {
+namespace sfm {
+
+/// Filter the toFilter iterable sequence into a new sequence without change the source sequence
+/// (keep only the element that share a common index with the provided Ids index list).
+/// See the reference: sfm_filters.hpp
+static void KeepOnlyReferencedElement(
+  const std::set<IndexT> & set_remainingIds,
+  const matching::PairWiseMatches & map_matches_in,
+  matching::PairWiseMatches & map_matches_out)
+{
+  matching::PairWiseMatches map_matches_E_infered;
+  for (const auto & iter_pairwise_matches : map_matches_in)
+  {
+    if (set_remainingIds.count(iter_pairwise_matches.first.first) &&
+      set_remainingIds.count(iter_pairwise_matches.first.second))
+    {
+      map_matches_E_infered.insert(iter_pairwise_matches);
+    }
+  }
+  map_matches_out.swap(map_matches_E_infered);
+}
+
+bool SplitMatchFileIntoMatchFiles(const SfM_Data & sfm_data, const std::string & match_file,
```

Comment:
Congrats for having the CI running ok.

Re-reading your code inspire me this. I think we can make the API more flexible for our users if we don't force the IO inside the function.

I think to this:
```
bool PairsToConnectedComponents(
  const Pair_Set & pairs,
  bool is_biedge,
  int min_nodes,
  std::map<IndexT, std::set<IndexT> > & subgraphs_ids);
// Then we can use the existing `KeepOnlyReferencedElement` function to will extract the subgraphs from the PairwiseMatches. KeepOnlyReferencedElement(const std::set<IndexT> & set_remainingIds, openMVG::matching::PairWiseMatches& map_matches)
```

So at the end you can use it on the fly, or to save the matches to the disk.

```
match IO
then call PairsToConnectedComponents
then call SplitMatchFileIntoMatches() -> that will output a std::vector<PairwiseMatches>
Then we can choose to log the matches to disk or use them directly.
```

Reply:
Yes, I will change the API.